core/core-impl/src/main/java/ai/timefold/solver/core/impl/score/director/AbstractScoreDirector.java

```diff
@@ -203,13 +211,16 @@ private void assertNonNullPlanningId(Object fact) {
 
     @Override
     public Score_ doAndProcessMove(Move<Solution_> move, boolean assertMoveScoreFromScratch) {
-        if (assertMoveScoreFromScratch) {
-            beforeMoveSnapshot = ShadowVariablesAssert.takeSnapshot(getSolutionDescriptor(), workingSolution);
+        if (solutionTracker != null) {
```

Comment:
I like the previous approach, where this condition was hidden behind a variable. The variable makes the purpose of the check obvious, while hiding what the check actually is.